src/Microsoft.Health.Fhir.Api/Modules/FhirModule.cs

```diff
@@ -57,12 +59,31 @@ public void Load(IServiceCollection services)
             services.AddSingleton(xmlParser);
             services.AddSingleton(xmlSerializer);
 
-            services.AddSingleton<IReadOnlyDictionary<ResourceFormat, Func<string, Resource>>>(x =>
+            ResourceElement SetMetadata(Resource resource, string versionId, DateTimeOffset lastModified)
             {
-                return new Dictionary<ResourceFormat, Func<string, Resource>>
+                resource.VersionId = versionId;
+                resource.Meta.LastUpdated = lastModified;
+                return resource.ToResourceElement();
+            }
+
+            services.AddSingleton<IReadOnlyDictionary<FhirResourceFormat, Func<string, string, DateTimeOffset, ResourceElement>>>(_ =>
+            {
+                return new Dictionary<FhirResourceFormat, Func<string, string, DateTimeOffset, ResourceElement>>
                 {
-                    { ResourceFormat.Json, str => jsonParser.Parse<Resource>(str) },
-                    { ResourceFormat.Xml, str => xmlParser.Parse<Resource>(str) },
+                    {
+                        FhirResourceFormat.Json, (str, version, lastModified) =>
+                        {
+                            var resource = jsonParser.Parse<Resource>(str);
+                            return SetMetadata(resource, version, lastModified);
```

Comment:
Just curious why this logic needs to happen in IoC container? It could be implemented in one place in `ResourceDeserializer.DeserializeRaw` right? Then the registration will much simpler just FhirResourceFormat against a method. I feel like having business logic injected in IoC container is a little odd?

Reply:
I'm not currently sure how to update properties on ITypedElement, ResourceDeserializer is still in Core so it doesn't have access to Resource. I don't have a good solution for this yet